xarray/core/common.py

```diff
@@ -568,9 +572,47 @@ def resample(self, freq, dim, how='mean', skipna=None, closed=None,
 
         .. [1] http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
         """
-        from .dataarray import DataArray
 
+        if dim is not None:
+            return self._resample_immediately(freq, dim, how, skipna, closed,
+                                              label, base, keep_attrs)
+
+        # More than one indexer is ambiguous, but we do in fact need one if
+        # "dim" was not provided, until the old API is fully deprecated
+        if len(indexer) != 1:
+            raise NotImplementedError(
+                "Resampling only supported along single dimensions."
+            )
+        dim, freq = indexer.popitem()
+
+        if isinstance(dim, basestring):
+            dim_name = dim
+            dim = self[dim]
+        else:
+            raise ValueError("Dimension name should be a string; "
+                             "was passed %r" % dim)
+
+        time_grouper = pd.TimeGrouper(freq=freq, closed=closed,
+                                      label=label, base=base)
+        resampler = self.resample_cls(self, group=dim, dim=dim_name,
+                                      grouper=time_grouper)
+
+        return resampler
+
+    def _resample_immediately(self, freq, dim, how, skipna,
+                              closed, label, base, keep_attrs):
+        """Implement the original version of .resample() which immediately
+        executes the desired resampling operation. """
+        from .dataarray import DataArray
         RESAMPLE_DIM = '__resample_dim__'
+
+        warnings.warn("\n.resample() has been modified to defer "
+                      "calculations. Instead of passing 'dim' and "
+                      "'how=\"{how}\", instead consider using "
+                      ".resample({dim}=\"{freq}\").{how}() ".format(
+                            dim=dim, freq=freq, how=how
+                      ), DeprecationWarning, stacklevel=3)
+
         if isinstance(dim, basestring):
             dim = self[dim]
         group = DataArray(dim, [(RESAMPLE_DIM, dim)], name=RESAMPLE_DIM)
```

Comment:
Note: I think this should probably be `group = DataArray(dim, [(dim.dims, dim)], name=RESAMPLE_DIM)` but for some reason this works.